src/backend/distributed/planner/multi_planner.c

```diff
@@ -117,38 +187,154 @@ multi_planner(Query *parse, int cursorOptions, ParamListInfo boundParams)
 
 
 /*
- * CreatePhysicalPlan encapsulates the logic needed to transform a particular
- * query into a physical plan. For modifications, queries immediately enter
- * the physical planning stage, since they are essentially "routed" to remote
- * target shards. SELECT queries go through the full logical plan/optimize/
- * physical plan process needed to produce distributed query plans.
+ * IsModifyCommand returns true if the query performs modifications, false
+ * otherwise.
  */
-MultiPlan *
-CreatePhysicalPlan(Query *originalQuery, Query *query,
-				   RelationRestrictionContext *restrictionContext)
+bool
+IsModifyCommand(Query *query)
 {
-	MultiPlan *physicalPlan = MultiRouterPlanCreate(originalQuery, query,
-													restrictionContext);
-	if (physicalPlan == NULL)
+	CmdType commandType = query->commandType;
+
+	if (commandType == CMD_INSERT || commandType == CMD_UPDATE ||
+		commandType == CMD_DELETE || query->hasModifyingCTE)
 	{
-		/* Create and optimize logical plan */
-		MultiTreeRoot *logicalPlan = MultiLogicalPlanCreate(query);
-		MultiLogicalPlanOptimize(logicalPlan);
+		return true;
+	}
+
+	return false;
+}
+
+
+/*
+ * CreateDistributedPlan creates the distributed plan for a query over
+ * distributed tables.
+ */
+static PlannedStmt *
+CreateDistributedPlan(PlannedStmt *localPlan, Query *originalQuery, Query *query,
+					  ParamListInfo boundParams,
+					  RelationRestrictionContext *restrictionContext)
+{
+	MultiPlan *physicalPlan = NULL;
+	PlannedStmt *resultPlan = NULL;
+	DeferredErrorMessage *error = NULL;
+	bool hasUnresolved = false;
+
+	if (HasUnresolvedExternParamsWalker((Node *) query, boundParams))
+	{
+		hasUnresolved = true;
+	}
+
+	if (IsModifyCommand(query))
+	{
+		/*
+		 * Modifications are always routed through the same executor. As
+		 * there's currently no other way to plan these, error out if the
+		 * query is unsupported.
+		 */
+		error = CreateModifyPlan(&physicalPlan, originalQuery, query,
+								 restrictionContext);
+
+		/*
+		 * Error out if there's an error, unless the error was possibly
+		 * triggered by missing parameters.  In that case an alternative plan
+		 * will be generated with bound parameters (see below).
+		 */
+		if (error && !hasUnresolved)
```

Comment:
It would be easier to understand what we're doing here if this read: `if (planningError && !suppressPlan)`